tools/PI/DevHome.PI/Controls/AddToolControl.xaml.cs

```diff
@@ -101,55 +161,290 @@ private void ClearValues()
             return null;
         }
 
-        var argType = ExternalToolArgType.None;
-
-        if (HwndRadio.IsChecked ?? false)
+        var activationType = ToolActivationType.Launch;
+        if (ProtocolRadio.IsChecked ?? false)
         {
-            argType = ExternalToolArgType.Hwnd;
+            activationType = ToolActivationType.Protocol;
         }
-        else if (ProcessIdRadio.IsChecked ?? false)
+        else if (selectedApp is not null && selectedApp.IsMsix)
         {
-            argType = ExternalToolArgType.ProcessId;
+            activationType = ToolActivationType.Msix;
         }
 
         return new(
             ToolNameTextBox.Text,
             ToolPathTextBox.Text,
-            argType,
-            PrefixTextBox.Text ?? string.Empty,
-            OtherArgsTextBox.Text ?? string.Empty,
+            activationType,
+            ArgumentsTextBox.Text,
+            selectedApp?.AppUserModelId ?? string.Empty,
+            selectedApp?.IconFilePath ?? string.Empty,
             IsPinnedToggleSwitch.IsOn);
     }
 
-    private void UpdateSampleCommandline(object sender, TextChangedEventArgs e)
+    private void CancelButton_Click(object sender, RoutedEventArgs e)
     {
-        UpdateSampleCommandline();
+        ClearValues();
     }
 
-    private void UpdateSampleCommandline(object sender, RoutedEventArgs e)
+    private async void AppListButton_Click(object sender, RoutedEventArgs e)
     {
-        UpdateSampleCommandline();
+        allApps.Clear();
+        SortedApps.Clear();
+
+        itemCount = GetShortcuts();
+        if (itemCount == 0)
+        {
+            Log.Error("Error getting shortcuts");
+        }
+
+        var packageCount = GetPackages();
+        if (packageCount == 0)
+        {
+            Log.Error("Error getting packages");
+        }
+
+        itemCount += packageCount;
+        if (itemCount == 0)
+        {
+            Log.Error("Error getting list of installed apps");
+            return;
+        }
+
+        // We get most of the data on a background thread, which
+        // reports intermittent progress.
+        IsLoading = true;
+        var progress = new Progress<int>(percent =>
+        {
+            // Update the progress report.
+            LoadingProgressTextRing.Value = percent;
+        });
+
+        await ProcessItemsAsync(progress);
+
+        foreach (var app in allApps)
+        {
+            SortedApps.Add(app);
+        }
+
+        IsLoading = false;
+    }
+
+    private int GetShortcuts()
```

Comment:
This method has a side-effect of populating `shortcuts`  Considering this returns `shortcuts.count` a more appropriate name would be `PopulateShortcuts` and return void.  The calling method can figure out the count.

Or, add a comment letting future developers know of the side-effect.

Reply:
We have both requirements: update the list and also return a count. Added comments in both GetShortcuts and GetPackages methods.